Lisk 1964 add hodlerdrop redistribution contract

hrmhatef opened the pull request:
### What was the problem?

This PR resolves [task](https://onchaincollective.atlassian.net/browse/LISK-1964)

### How was it solved?

Add a new contract to solve the problem

### How was it tested?

Add some new unit tests for the new contract.

hrmhatef committed:
:seedling: Add a new contract for the HodlerdropV2

hrmhatef committed:
:seedling: Add a new unit test for the HodlerdropV2

hrmhatef committed:
:recycle: Update the contract

hrmhatef committed:
:recycle: Add some additional tests

hrmhatef committed:
:recycle: Add claimed tests

hrmhatef committed:
:recycle: Fix a bug and update some messages

hrmhatef committed:
:recycle: Update the unit test

hrmhatef committed:
:recycle: Update the docs

Phanco commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+        );
+        require(_ecosystemFundAddress != address(0), "L2AirdropV2: Ecosystem Fund wallet address can not be zero");
+        l2LiskTokenAddress = _l2LiskTokenAddress;
+        l2LockingPositionAddress = _l2LockingPositionAddress;
```
Swap this line and the line below so that these 4 params are in order

Phanco commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+    /// @param liskAddress The Lisk v4 address to check if it has claimed the full airdrop.
+    /// @return True if the Lisk v4 address has claimed the full airdrop, False otherwise.
+    function claimedFullAirdrop(bytes20 liskAddress) public view returns (bool) {
+        return (airdropStatus[liskAddress] & FULL_AIRDROP_CLAIMED) == FULL_AIRDROP_CLAIMED;
```
Can we use `airdropStatus[liskAddress] == FULL_AIRDROP_CLAIMED` instead (I know this approach is referenced from L2Airdrop.sol, just curious)

matjazv replied:
I would recommend not modifying any lines if not strictly necessary then what's inside `L2Airdrop.sol`, as this contract has been audited by a reputable firm and thoroughly battle-tested. It should be used as our baseline.

hrmhatef replied:
It's the same with the original one.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
Maybe we could have a more descriptive file name like `L2HodlerdropRedistribution.sol`

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+import { IL2VotingPower } from "../interfaces/L2/IL2VotingPower.sol";
+
+/// @title L2AirdropV2
+/// @notice L2AirdropV2 is an implementation of the Lisk v4 migration airdropV2 on L2 version 2.
```
```suggestion
/// @notice L2AirdropV2 distributes the remaining LSK tokens of the Hodlerdrop.
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+
+/// @title L2AirdropV2
+/// @notice L2AirdropV2 is an implementation of the Lisk v4 migration airdropV2 on L2 version 2.
+///         It is responsible for the airdropV2
```
I would remove this line

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+///         The claim amount for each user is equal to their share in the remaining unclaimed balance of
+///         the L2Airdrop contract,
+///         where their share is calculated proportionately to the amount they already claimed.
+///         computation is based on the following conditions:
```
```suggestion
///         Computation is based on the following conditions:
```

matjazv replied:
Perhaps change documentation to something like below (pay attention to text alignment)
```solidity
/// @notice L2AirdropV2 is an implementation of the Lisk v4 migration airdrop on L2 version 2.
///         It distributes LSK tokens to eligible users based on their previous claims in L2Airdrop.
///         Each user's claim amount is proportional to their share of the total claimed amount in L2Airdrop,
///         applied to the remaining unclaimed balance. The distribution eligibility is determined by the
///         following staking conditions:
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+///         1. Staking Tier 1: A user is required to stake the same amount they received
+///                            in the Hodlerdrop for a period of time (3 months).
+///         2. Staking Tier 2: This condition is analogous to the previous, but a user is required
+///                            to stake the same amount for a twice longer period of time (6months).
```
```suggestion
///                            to stake the same amount for twice as long (6months).
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+///         conditions.
+///         Any remaining amount left in the contract will be moved to the Ecosystem Fund.
+contract L2AirdropV2 is Ownable2Step {
+    /// @notice The total amount of LSK tokens to be given away in the migration airdrop.
```
```suggestion
    /// @notice The total amount of LSK tokens to be given away in the hodlerdrop.
```

AndreasKendziorra replied:
I know it was copied from confluence. But here we should try to use consistent wording, i.e., either "Hodlerdrop" or "migration airdrop".

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+///         Any remaining amount left in the contract will be moved to the Ecosystem Fund.
+contract L2AirdropV2 is Ownable2Step {
+    /// @notice The total amount of LSK tokens to be given away in the migration airdrop.
+    uint256 public constant MIGRATION_AIRDROP_AMOUNT = 3000000 * 10 ** 18;
```
Ideally, we also rename 
- `MIGRATION_AIRDROP_AMOUNT` to `HODLERDROP_AMOUNT`
- `MIGRATION_AIRDROP_UNCLAIMED_AMOUNT` to `HODLERDROP_UNCLAIMED_AMOUNT`
- `MIGRATION_AIRDROP_CLAIMED_AMOUNT` to `HODLERDROP_CLAIMED_AMOUNT`

hrmhatef replied:
since we don't need them any more, I removed them

hrmhatef committed:
:recycle: Apply some comments

matjazv commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,282 @@
+
+    /// @notice The total amount of LSK tokens claimed during HodlerdropV2.
+    uint256 public constant MIGRATION_AIRDROP_CLAIMED_AMOUNT =
+        MIGRATION_AIRDROP_AMOUNT - MIGRATION_AIRDROP_UNCLAIMED_AMOUNT;
```
Why do we need these constants inside smart contract? They are not used anywhere.

matjazv replied:
If we need to have them, then add unit test to check their values.
Something like this:
```solidity
function test_MigrationConstants() public {
    // verify the migration constants are correctly set
    assertEq(l2AirdropV2.MIGRATION_AIRDROP_AMOUNT(), 3000000 * 10 ** 18);
    assertEq(l2AirdropV2.MIGRATION_AIRDROP_UNCLAIMED_AMOUNT(), 1375134441061135000000000);
    assertEq(l2AirdropV2.MIGRATION_AIRDROP_CLAIMED_AMOUNT(), 
        l2AirdropV2.MIGRATION_AIRDROP_AMOUNT() - l2AirdropV2.MIGRATION_AIRDROP_UNCLAIMED_AMOUNT());
}
```

hrmhatef replied:
I just add them based on the contract Specification.
Maybe @AndreasKendziorra can help us here, Do we need them in the contract or not?

AndreasKendziorra replied:
These constants are not needed here. The specs on Confluence are for whole process, and these constants are required for creating the Merkle tree, but not for the contract.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    /// @notice The period of time starting from the setting of the Merkle root, during which the airdrop can be
+    ///         claimed.
+    ///         All LSK tokens should be transferable to the Lisk Ecosystem funds afterwards.
+    uint32 public constant HODLER_AIRDROPV2_DURATION = 180; // 6 months
```
This one could be renamed to `HODLERDROP_REDISTRIBUTION_DURATION` to better distinguish it from the hodlerdrop

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    /// @notice Start time of the migration airdrop. AirdropV2 is considered started once the Merkle root is set.
+    uint256 public airdropStartTime;
+
+    /// @notice Mapping of the airdropV2 status for each Lisk v4 address. In particular, for each of the airdrop
```
```suggestion
    /// @notice Mapping of the airdropV2 status for each L2 address. In particular, for each of the airdrop
```

AndreasKendziorra replied:
alternatively, we could also say "for each recipient address"

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+
+    /// @notice Mapping of the airdropV2 status for each Lisk v4 address. In particular, for each of the airdrop
+    ///         conditions (staking tier 1, staking tier 2).
+    mapping(bytes20 => uint8) public airdropStatus;
```
since we are using Ethereum addresses, we should use the type `address` instead of `bytes20`.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    event LSKSentToEcosystemWallet(address indexed ecosystemWalletAddress, uint256 amount);
+
+    /// @notice Emitted when the airdrop is (partially) claimed for the recipient.
+    event AirdropClaimed(bytes20 indexed liskAddress, uint256 amount, address indexed recipient, uint8 airdropStatus);
```
the argument `liskAddress` is redundant and should be removed. Only `recipient` matters for this airdrop.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    /// @param recipient The recipient address to check if it satisfies the staking requirement of the provided tier.
+    /// @param airdropAmount The amount of LSK tokens to claim the airdrop for.
+    /// @param tierDuration The duration of the staking requirement for the provided tier.
+    /// @return True if recipient has staked at least minStakingAmount for at least MIN_STAKING_DURATION_TIER_1 or
```
```suggestion
    /// @return True if recipient has staked at least airdropAmount for at least MIN_STAKING_DURATION_TIER_1 or
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+        IL2LockingPosition.LockingPosition[] memory lockingPositions =
+            l2LockingPosition.getAllLockingPositionsByOwner(recipient);
+
+        // check if the recipient has staked at least minStakingAmount for at least tierDuration
```
```suggestion
        // check if the recipient has staked at least airdropAmount for at least tierDuration
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    /// @notice Check if the Lisk v4 address has claimed the airdrop for staking tier 1.
+    /// @param liskAddress The Lisk v4 address to check if it has claimed the airdrop for staking tier 1.
+    /// @return True if the Lisk v4 address has claimed the airdrop for staking tier 1, False otherwise.
+    function claimedStakingTier1(bytes20 liskAddress) public view returns (bool) {
```
We don't deal with Lisk v4 addresses here. Only Ethereum addresses. More specifically, all recipient addresses of the hodler drop. So, the argument should be of type `address` and could be called `recipient` or `recipientAddress`. The lines before must also be adapted.

The same for `claimedStakingTier2` and `claimedFullAirdrop`.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    /// @notice Check if the recipient satisfies the staking requirement of tier 1.
+    /// @param recipient The recipient address to check if it satisfies the staking requirement of tier 1.
```
I know that it was just copied. But the parameter `airdropAmount` is missing here. You could just copy the text from the function `satisfiesStakingTier`.

The same for `satisfiesStakingTier2`.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+
+    /// @notice Check if the recipient satisfies the staking requirement of tier 1.
+    /// @param recipient The recipient address to check if it satisfies the staking requirement of tier 1.
+    /// @return True if recipient has staked at least minStakingAmount for at least MIN_STAKING_DURATION_TIER_1, False
```
```suggestion
    /// @return True if recipient has staked at least airdropAmount for at least MIN_STAKING_DURATION_TIER_1, False
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+
+    /// @notice Check if the recipient satisfies the staking requirement of tier 2.
+    /// @param recipient The recipient address to check if it satisfies the staking requirement of tier 2.
+    /// @return True if recipient has staked at least minStakingAmount for at least MIN_STAKING_DURATION_TIER_2, False
```
```suggestion
    /// @return True if recipient has staked at least airdropAmount for at least MIN_STAKING_DURATION_TIER_2, False
```

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    /// @param liskAddress The Lisk v4 address to claim the airdrop for.
+    /// @param amount The amount of LSK tokens to claim the airdrop for.
+    /// @param merkleProof The Merkle proof for the liskAddress and amount against the stored merkleRoot.
+    function claimAirdrop(bytes20 liskAddress, uint256 amount, bytes32[] memory merkleProof) public {
```
Same here: It should be an Ethereum address, not a Lisk v4 address. It could be called `recipient` or `recipientAddress`.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+        require(
+            IL2Claim(l2ClaimAddress).claimedTo(liskAddress) != address(0),
+            "L2AirdropV2: tokens were not claimed yet from this Lisk address"
+        );
```
This check is not meaningful for this contract and should be removed. The reason is that we don't care about Lisk addresses. We only care about Ethereum addresses, namely the recipient addresses of the hodler drop. And we ensure that a provided address was indeed a recipient in the hodlerdrop via the Merkle proof.

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+        );
+        require(amount > 0, "L2AirdropV2: amount is zero");
+        require(merkleProof.length > 0, "L2AirdropV2: Merkle proof is empty");
+        // require merkleProof be a correct proof for liskv4Address and amount against stored merkleRoot
```
also in this comment, the name of the variable `liskv4Address` must be change to `recipient`, `recipientAddress` or what ever you choose

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+            "L2AirdropV2: full airdrop claimed"
+        );
+
+        address recipient = IL2Claim(l2ClaimAddress).claimedTo(liskAddress);
```
this line is redundant because the recipient address must be provided as an argument

AndreasKendziorra commented on src/L2/L2AirdropV2.sol:
```diff
@@ -0,0 +1,281 @@
+    address public immutable l2LiskTokenAddress;
+
+    /// @notice Address of the L2Claim contract.
+    address public immutable l2ClaimAddress;
```
This address is not needed here and should be removed. See also the other comments for hte lines were this variable is used.

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+    L2LiskToken public l2LiskToken;
+    address public remoteToken;
+    address public bridge;
+    L2Claim public l2Claim;
```
anything regarding l2claim should be redundant in this file and should be removed.

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+
+        // TODO make sure we dont need this part
+        // initialize Lisk DAO Treasury contract inside L2Staking contract
+        //l2Staking.initializeDaoTreasury(ecosystemFundWalletAddress);
```
for the correct setup of the l2Staking contract, this function must be called. I guess the tests here would work also without calling this function. But just in case, I would do it. 
The address you provide as an argument is not very important. The staking contract is transferring the penalties for "fast unlocks" to this address, and in the real setup, the recipient is the treasury of the DAO. So for here, it's fine to use `ecosystemFundWalletAddress` or any other address.

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+        assertEq(l2Claim.claimedTo(aliceLSKAddress), alice);
+    }
+
+    function test_MigrationConstants() public view {
```
as discussed above, these constants are not need

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+
+    address ecosystemFundWalletAddress;
+    address alice;
+    bytes20 aliceLSKAddress;
```
not needed

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+        bytes32[] memory merkleProof = new bytes32[](1);
+        merkleProof[0] = bytes32(0xf0df3dcda05b4fbd9c655cde3d5ceb211e019e72ec816e127a59e7195f2cd7f5);
+        l2AirdropV2.claimAirdrop(aliceLSKAddress, 80 * 10 ** 18, merkleProof);
+        assertEq(l2LiskToken.balanceOf(alice), aliceBalanceBefore + 40 * 10 ** 18); // 4 L2LiskToken airdrop
```
it's not clear to me what the comment in the end of the line means. Probably it must adjusted to "40" to match with the initial intention. But since it's unclear it may be better to remove it.

The same in aliceClaimAirdropForStakingTier2.

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+        aliceClaimAirdropForStakingTier2();
+    }
+
+    function test_ClaimAirdrop_StakingTier1_StakingTier2() public {
```
this test could also be removed as everything is repeated in test_ClaimAirdrop_FullAirdrop

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+        );
+
+        l2AirdropV2.claimAirdrop(aliceLSKAddress, 80 * 10 ** 18, merkleProof);
+        assertEq(l2LiskToken.balanceOf(alice), aliceBalanceBefore + 80 * 10 ** 18); // 16 L2LiskToken airdrop
```
either remove the comment or change 16 to 80 if you think it makes sense (for me, the meaning is unclear)

AndreasKendziorra commented on test/L2/L2AirdropV2.t.sol:
```diff
@@ -0,0 +1,594 @@
+        l2AirdropV2.claimAirdrop(bytes20(alice), 20 * 10 ** 18, merkleProof);
+    }
+
+    function test_ClaimAirdrop_ZeroRecipientAddress() public {
```
this test should be removed, because the eligibility should only be checked via the Merkle tree, not the claims contract.

hrmhatef replied:
We need to have this test to check the recipient address to be not zero.

AndreasKendziorra replied:
You are right. Now that the recipient address is provided, this check makes sense.

AndreasKendziorra reviewed (changes requested):
The script for deploying, similar to https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol, is missing.

matjazv commented:
> The script for deploying, similar to https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol, is missing.

The deployment shell script is also missing, similar to the one here:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/5_deployAirdropContract.sh

hrmhatef committed:
:recycle: Rename L2Airdrop -> L2HodlerdropRedistribution
 - Also edit the contract to use address instead of lisk address
 - Update the unit test too

hrmhatef committed:
:seedling: Add a new script to deploy the L2HodlerdropRedistribution

Phanco commented:
If the file is renamed to `L2HodlerdropRedistribution.sol`, should the contract name also be renamed to `L2HodlerdropRedistribution`?

hrmhatef committed:
:recycle: Remove unused test

hrmhatef committed:
:recycle: Rename airdrop -> hodlerdrop

hrmhatef committed:
:recycle: Update the utils for the Hodlerdrop address
 - Also edit the script based on the last rename

hrmhatef committed:
:recycle: Add a new unit test for 400 locking positions

matjazv commented:
> > The script for deploying, similar to https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol, is missing.
> 
> The deployment shell script is also missing, similar to the one here: https://github.com/LiskHQ/lisk-contracts/blob/development/script/5_deployAirdropContract.sh

Shell script is still missing.

matjazv commented on script/contracts/Utils.sol:
```diff
@@ -23,6 +23,8 @@ contract Utils is Script {
         /// @notice L2 Airdrop address.
         address L2Airdrop;
+        /// @notice L2 Hodlerdrop Redistribution address.
+        address L2HodlerdropRedistribution;
```
Variables inside this struct must be lexicographically ordered.

matjazv commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+        console2.log("L2 Locking Position address: %s", l2AddressesConfig.L2LockingPosition);
+
+        // get Merkle root
+        Utils.MerkleRoot memory merkleRoot = utils.readMerkleRootFile("hodlerdrop-merkle-root.json");
```
Do we already have any (test) merkle root available?
We have them in place for 1st Airdrop here:
https://github.com/LiskHQ/lisk-contracts/tree/LISK-1964-add-hodlerdrop-redistribution-contract/script/data
(inside `devnet`, `testnet`, `mainnet` folders)

hrmhatef replied:
@Phanco is working on it to prepare the tree-builder, after that I can add this part too.

matjazv commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+        );
+
+        // get L2 Ecosystem Fund wallet address where LSK tokens will be transferred after Hodlerdrop period is over
+        address ecosystemFundWalletAddress = vm.envAddress("L2_ECOSYSTEM_FUND_WALLET_ADDRESS");
```
This variable needs to be set inside `.env.devnet`, `.env.testnet` and `.env.mainnet` files.
`.env.testnet` file:
https://github.com/LiskHQ/lisk-contracts/blob/LISK-1964-add-hodlerdrop-redistribution-contract/.env.testnet

matjazv commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+        console2.log("Deploying L2 HodlerdropRedistribution contract...");
+
+        // address, the ownership of L2 HodlerdropRedistribution contract is transferred to after deployment
+        address newOwnerAddress = vm.envAddress("L2_HODLERDROP_REDISTRIBUTION_OWNER_ADDRESS");
```
This variable needs to be set inside `.env.devnet`, `.env.testnet` and `.env.mainnet` files.
`.env.testnet` file:
https://github.com/LiskHQ/lisk-contracts/blob/LISK-1964-add-hodlerdrop-redistribution-contract/.env.testnet

matjazv commented on test/utils/Utils.t.sol:
```diff
@@ -38,6 +38,7 @@ contract UtilsTest is Test {
         uint160 index = 1;
         Utils.L2AddressesConfig memory config = Utils.L2AddressesConfig({
             L2Airdrop: address(index++),
+            L2HodlerdropRedistribution: address(index++),
```
Set this variable to be lexicographically ordered.

hrmhatef committed:
:recycle: Change the order of the new variables

hrmhatef committed:
:recycle: Add new variables into the env files

hrmhatef committed:
:seedling: Add hodlerdrop tree info for all networks

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -6,73 +6,60 @@ import { Ownable2Step } from "@openzeppelin/contracts/access/Ownable2Step.sol";
-///         each recipient is stored in a mapping. The airdrop status includes the status of each of the airdrop
-///         conditions.
+///                            to stake the same amount for twice as long (6months).
+///         The airdrop amount is distributed to the recipient's address in L2LiskToken contract. The hodlerdrop status
```
I see that (almost?) everywhere "airdrop" was replaced by "hodlerdrop". But this could quite confusing because the airdrop we are talking about is the "hodlerdrop redistribution", not the "hodlerdrop". Therefore, I would either revert it (using the more general term "airdrop" is fine) or change it to "hodlerdrop redistribution".

hrmhatef replied:
I changed them to `hodlerdrop-redistribution`, WDYT?

AndreasKendziorra replied:
yes, sounds good to me

matjazv commented on script/contracts/Utils.sol:
```diff
@@ -198,6 +200,11 @@ contract Utils is Script {
             l2AddressesConfig.L2Airdrop = l2Airdrop;
         } catch { }
 
+        try vm.parseJsonAddress(addressJson, ".L2HodlerdropRedistribution") returns (address l2HodlerdropRedistribution)
+        {
+            l2AddressesConfig.L2HodlerdropRedistribution = l2HodlerdropRedistribution;
+        } catch { }
+
```
For consistency, move these lines below `L2GovernorPaused`.

matjazv commented on script/contracts/Utils.sol:
```diff
@@ -335,6 +342,7 @@ contract Utils is Script {
     function writeL2AddressesFile(L2AddressesConfig memory cfg, string memory filePath) external {
         string memory json = "";
         vm.serializeAddress(json, "L2Airdrop", cfg.L2Airdrop);
+        vm.serializeAddress(json, "L2HodlerdropRedistribution", cfg.L2HodlerdropRedistribution);
```
For consistency, move these lines below `L2GovernorPaused`.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
+    /// @notice Claim the Hodlerdrop for the recipient.
+    /// @param recipient The recipient address to claim the hodlerdrop for.
+    /// @param amount The amount of LSK tokens to claim the hodlerdrop for.
+    /// @param merkleProof The Merkle proof for the address and the amount against the stored merkleRoot.
```
```suggestion
    /// @param merkleProof The Merkle proof for the recipient address and the amount against the stored merkleRoot.
```

AndreasKendziorra replied:
or just "for recipient"

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
+        require(recipient != address(0), "L2HodlerdropRedistribution: recipient is the zero address");
+        require(amount > 0, "L2HodlerdropRedistribution: amount is zero");
+        require(merkleProof.length > 0, "L2HodlerdropRedistribution: Merkle proof is empty");
+        // require merkleProof be a correct proof for address and amount against stored merkleRoot
```
```suggestion
        // require merkleProof be a correct proof for the recipient address and amount against stored merkleRoot
```

AndreasKendziorra replied:
or just "for recipient"

AndreasKendziorra commented on test/L2/L2HodlerdropRedistribution.t.sol:
```diff
@@ -4,39 +4,36 @@ pragma solidity 0.8.23;
 import { Test, console2, stdStorage, StdStorage } from "forge-std/Test.sol";
-import { L2AirdropV2 } from "src/L2/L2AirdropV2.sol";
+import { L2HodlerdropRedistribution } from "src/L2/L2HodlerdropRedistribution.sol";
 import { L2Claim } from "src/L2/L2Claim.sol";
```
this import can be removed

AndreasKendziorra commented on .env.devnet:
```diff
@@ -37,6 +37,12 @@ L2_VESTING_WALLET_OWNER_ADDRESS=0x
+# Owner of L2HodlerdropRedistribution contract
+L2_HODLERDROP_REDISTRIBUTION_OWNER_ADDRESS=0x
+
+# EcosystemFund wallet address where LSK tokens will be transferred after Hodlerdrop period is over
```
```suggestion
# EcosystemFund wallet address where LSK tokens will be transferred to after Hodlerdrop redistribution period is over
```

AndreasKendziorra commented on .env.mainnet:
```diff
@@ -37,6 +37,12 @@ L2_AIRDROP_OWNER_ADDRESS=0x394Ae9d48eeca1C69a989B5A8C787081595c55A7
+# Owner of L2HodlerdropRedistribution contract
+L2_HODLERDROP_REDISTRIBUTION_OWNER_ADDRESS=0x394Ae9d48eeca1C69a989B5A8C787081595c55A7
+
+# EcosystemFund wallet address where LSK tokens will be transferred after Hodlerdrop period is over
```
```suggestion
# EcosystemFund wallet address where LSK tokens will be transferred to after Hodlerdrop redistribution period is over
```

AndreasKendziorra commented on .env.testnet:
```diff
@@ -40,6 +40,12 @@ L2_AIRDROP_WALLET_ADDRESS=0xaF3f4B5033Ef666b3F3EA91a0f360541cC21e5d7
+# Owner of L2HodlerdropRedistribution contract
+L2_HODLERDROP_REDISTRIBUTION_OWNER_ADDRESS=0x84277C9255b3B6904aCfe12C2c45c6a36E81B059
+
+# EcosystemFund wallet address where LSK tokens will be transferred after Hodlerdrop period is over
```
```suggestion
# EcosystemFund wallet address where LSK tokens will be transferred to after Hodlerdrop redistribution period is over
```

AndreasKendziorra commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+
+/// @title L2AirdropScript - L2 Hodlerdrop Redistribution deployment script
+/// @notice This contract is used to deploy L2 HodlerdropRedistribution contract.
+contract L2AirdropScript is Script {
```
not strictly necessary, but it may better to rename it to `L2HodlerdropRedistributionScript`

AndreasKendziorra commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+            "L2 HodlerdropRedistribution owner address: %s (after ownership will be accepted)", newOwnerAddress
+        );
+
+        // get L2 Ecosystem Fund wallet address where LSK tokens will be transferred after Hodlerdrop period is over
```
```suggestion
        // get L2 Ecosystem Fund wallet address where LSK tokens will be transferred to after Hodlerdrop redistribution period is over
```

AndreasKendziorra commented on script/contracts/L2/L2HodlerdropRedistribution.s.sol:
```diff
@@ -0,0 +1,76 @@
+        console2.log("L2 Locking Position address: %s", l2AddressesConfig.L2LockingPosition);
+
+        // get Merkle root
+        Utils.MerkleRoot memory merkleRoot = utils.readMerkleRootFile("hodlerdrop-merkle-root.json");
```
also this json file should be renamed to something like `hodlerdrop-redistribution-merkle-root.json` to clearly distinguish it from the hodlerdrop

AndreasKendziorra commented on script/data/devnet/hodlerdrop-merkle-tree-result.json:
do we need this file here? or any of the other Merkle tree files (except for the Merkle root files)?

hrmhatef replied:
Based on the airdrop we don't need this file.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
-    /// @return True if the Lisk v4 address has claimed the airdrop for staking tier 1, False otherwise.
-    function claimedStakingTier1(bytes20 liskAddress) public view returns (bool) {
-        return (airdropStatus[liskAddress] & STAKING_TIER_1_BIT) != 0;
+    /// @notice Check if the address has claimed the hodlerdrop for staking tier 1.
```
```suggestion
    /// @notice Check if the recipient address has claimed the hodlerdrop for staking tier 1.
```

AndreasKendziorra replied:
or just "the recipient". The same for the other lines

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
-    /// @return True if the Lisk v4 address has claimed the airdrop for staking tier 2, False otherwise.
-    function claimedStakingTier2(bytes20 liskAddress) public view returns (bool) {
-        return (airdropStatus[liskAddress] & STAKING_TIER_2_BIT) != 0;
+    /// @notice Check if the address has claimed the hodlerdrop for staking tier 2.
```
```suggestion
    /// @notice Check if the recipient address has claimed the hodlerdrop for staking tier 2.
```

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
-    /// @return True if the Lisk v4 address has claimed the full airdrop, False otherwise.
-    function claimedFullAirdrop(bytes20 liskAddress) public view returns (bool) {
-        return (airdropStatus[liskAddress] & FULL_AIRDROP_CLAIMED) == FULL_AIRDROP_CLAIMED;
+    /// @notice Check if the address has claimed the full hodlerdrop.
```
```suggestion
    /// @notice Check if the recipient address has claimed the full hodlerdrop.
```

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
-        return (airdropStatus[liskAddress] & STAKING_TIER_2_BIT) != 0;
+    /// @notice Check if the address has claimed the hodlerdrop for staking tier 2.
+    /// @param recipient The address to check if it has claimed the hodlerdrop for staking tier 2.
+    /// @return True if the address has claimed the hodlerdrop for staking tier 2, False otherwise.
```
```suggestion
    /// @return True if the recipient address has claimed the hodlerdrop for staking tier 2, False otherwise.
```

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -151,131 +141,129 @@ contract L2AirdropV2 is Ownable2Step {
-        return (airdropStatus[liskAddress] & FULL_AIRDROP_CLAIMED) == FULL_AIRDROP_CLAIMED;
+    /// @notice Check if the address has claimed the full hodlerdrop.
+    /// @param recipient The address to check if it has claimed the full hodlerdrop.
+    /// @return True if the address has claimed the full hodlerdrop, False otherwise.
```
```suggestion
    /// @return True if the recipient address has claimed the full hodlerdrop, False otherwise.
```

hrmhatef committed:
:recycle: Apply some comments

matjazv commented:
> > > The script for deploying, similar to https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol, is missing.
> > 
> > 
> > The deployment shell script is also missing, similar to the one here: https://github.com/LiskHQ/lisk-contracts/blob/development/script/5_deployAirdropContract.sh
> 
> Shell script is still missing.

Script (`.sh`) is still missing.

After you add it, please execute it to deploy the smart contract on the Lisk Sepolia network so that we can verify everything is working as expected.

AndreasKendziorra commented on test/L2/L2HodlerdropRedistribution.t.sol:
```diff
@@ -157,145 +144,136 @@ contract L2AirdropV2Test is Test {
 
+    function test_charlieSatifiesStakingTier1With400LockingPositions() public {
+        vm.startPrank(charlie);
+        // charlie stakes 1 L2LiskToken for minimum days in 400 positions
```
I think the initial intention to test such a large number was to ensure that a claim transaction doesn't run out of gas, right?
Therefore, it would be good to have a test that calls the claim function for this setup and also checks that the block gas limit was not exceeded. This could be done by the `gasleft()` function. 
Moreover, since we had already the case of 307 positions in the hodlerdrop, I would tend to choose an even larger number to ensure it will work for every user.

hrmhatef replied:
Increase the number of locking positions to 1000, also check the `gasleft()`.

hrmhatef committed:
:recycle: Increase the number of locking positions to 1000

hrmhatef committed:
:seedling: Add shell script for the L2HodlerdropRedistribution contract

hrmhatef committed:
:recycle: Refactor the gas usage for the claim function

matjazv commented on test/L2/L2HodlerdropRedistribution.t.sol:
```diff
@@ -0,0 +1,566 @@
+        merkleProof[1] = bytes32(0xf0df3dcda05b4fbd9c655cde3d5ceb211e019e72ec816e127a59e7195f2cd7f5);
+        l2HodlerdropRedistribution.claimHodlerdrop(charlie, 1000 * 10 ** 18, merkleProof);
+        assertEq(l2LiskToken.balanceOf(charlie), charlieBalanceBefore + 500 * 10 ** 18);
+        assertGe(gasleft(), 100);
```
Could we also add `gasleft()` at the beginning of the unit test and calculate the gas difference at the end?
The difference should not exceed 30 million, as this is the current block gas limit on Lisk.

matjazv replied:
If we are left with `100` in the current unit tests, does this mean that `1000` locking positions would be our maximum?

hrmhatef replied:
refactor the test to check the gas for the claim func to be less than our gas limit per block.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -29,28 +29,28 @@ contract L2HodlerdropRedistribution is Ownable2Step {
-    ///         set.
+    /// @notice Start time of the hodlerdrop-redistribution. L2HodlerdropRedistribution is considered started once the
+    ///         Merkle root is set.
     uint256 public hodlerdropStartTime;
```
I would also rename this one to `hodlerdropRedistributionStartTime`. Or if it's too long, then just `startTime`.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -29,28 +29,28 @@ contract L2HodlerdropRedistribution is Ownable2Step {
-    ///         conditions (staking tier 1, staking tier 2).
+    /// @notice Mapping of the hodlerdrop-redistribution status for each recipient address. In particular, for each of
+    ///         the hodlerdrop-redistribution conditions (staking tier 1, staking tier 2).
     mapping(address => uint8) public hodlerdropStatus;
```
Same here: Either `hodlerdropRedistributionStatus` or `status`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -29,28 +29,28 @@ contract L2HodlerdropRedistribution is Ownable2Step {
     uint8 public constant STAKING_TIER_2_BIT = 0x02;
-    // full hodlerdrop claimed
+    // full hodlerdrop-redistribution claimed
     uint8 public constant FULL_HODLERDROP_CLAIMED = 0x03;
```
Also here, it would be better to use `FULL_HODLERDROP_REDISTRIBUTION_CLAIMED`. Alternatively, `FULL_AIRDROP_CLAIMED`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -69,7 +69,7 @@ contract L2HodlerdropRedistribution is Ownable2Step {
 
-    /// @notice Emitted when the Hodlerdrop is (partially) claimed for the recipient.
+    /// @notice Emitted when the Hodlerdrop-redistribution is (partially) claimed for the recipient.
     event HodlerdropClaimed(uint256 amount, address indexed recipient, uint8 hodlerdropStatus);
```
would better to use `HodlerdropRedistributionClaimed`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -102,7 +102,7 @@ contract L2HodlerdropRedistribution is Ownable2Step {
     /// @notice Check if the recipient satisfies the staking requirement of the provided tier.
     /// @param recipient The recipient address to check if it satisfies the staking requirement of the provided tier.
-    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop for.
+    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop-redistribution for.
```
would be good to rename `hodlerdropAmount` to `hodlerdropRedistributionAmount` or `claimableAmount`. If you don't use `hodlerdropRedistributionAmount`, the error message in line 119 must probably be adapted, e.g. to "claimable amount is zero", depending on which variable name you choose.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -172,30 +172,32 @@ contract L2HodlerdropRedistribution is Ownable2Step {
+    /// @notice Check if the recipient address has claimed the full hodlerdrop-redistribution.
+    /// @param recipient The address to check if it has claimed the full hodlerdrop-redistribution.
+    /// @return True if the recipient address has claimed the full hodlerdrop-redistribution, False otherwise.
     function claimedFullHodlerdrop(address recipient) public view returns (bool) {
```
would be better to call it `claimedFullHodlerdropRedistribution`.

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -172,30 +172,32 @@ contract L2HodlerdropRedistribution is Ownable2Step {
     /// @notice Check if the recipient satisfies the staking requirement of tier 1.
     /// @param recipient The recipient address to check if it satisfies the staking requirement of tier 1.
-    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop for.
+    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop-redistribution for.
```
same as above: `hodlerdropRedistributionAmount` or `claimableAmount`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -204,27 +206,27 @@ contract L2HodlerdropRedistribution is Ownable2Step {
     /// @notice Check if the recipient satisfies the staking requirement of tier 2.
     /// @param recipient The recipient address to check if it satisfies the staking requirement of tier 2.
-    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop for.
+    /// @param hodlerdropAmount The amount of LSK tokens to claim the hodlerdrop-redistribution for.
```
same as above: `hodlerdropRedistributionAmount` or `claimableAmount`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -204,27 +206,27 @@ contract L2HodlerdropRedistribution is Ownable2Step {
         return satisfiesStakingTier(recipient, hodlerdropAmount, MIN_STAKING_DURATION_TIER_2);
     }
 
     /// @notice Claim the Hodlerdrop for the recipient.
```
```suggestion
    /// @notice Claim the Hodlerdrop redistribution for the recipient.
```

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -204,27 +206,27 @@ contract L2HodlerdropRedistribution is Ownable2Step {
+    /// @param recipient The recipient address to claim the hodlerdrop-redistribution for.
+    /// @param amount The amount of LSK tokens to claim the hodlerdrop-redistribution for.
+    /// @param merkleProof The Merkle proof for the recipient address and the amount against the stored merkleRoot.
     function claimHodlerdrop(address recipient, uint256 amount, bytes32[] memory merkleProof) public {
```
would be better to call it `claimHodlerdropRedistribution` or just `claim`

AndreasKendziorra commented on src/L2/L2HodlerdropRedistribution.sol:
```diff
@@ -233,7 +235,7 @@ contract L2HodlerdropRedistribution is Ownable2Step {
+            "L2HodlerdropRedistribution: full hodlerdrop-redistribution claimed"
         );
 
         uint256 hodlerdropAmount = 0;
```
would be better to use  `hodlerdropRedistributionAmount`

hrmhatef committed:
:recycle: Edit to apply some suggestions

matjazv commented:
May you execute `15_deployHodlerdropRedistributionContract.sh` script and deploy the smart contract on Lisk Sepolia that we check and test everything is working? You may use your private address(es) for `L2_HODLERDROP_REDISTRIBUTION_OWNER_ADDRESS` and `L2_ECOSYSTEM_FUND_WALLET_ADDRESS`.
